control-plane/pkg/reconciler/consumergroup/consumergroup.go

```diff
@@ -55,71 +71,87 @@ func (r Reconciler) reconcileConsumers(ctx context.Context, cg *kafkainternals.C
 		return cg.MarkReconcileConsumersFailed("ListConsumers", err)
 	}
 
-	// Stable sort consumers so that we give consumers different deletion
-	// priorities based on their state.
-	//
-	// Consumers at the tail of the list are deleted first.
-	sort.SliceStable(existingConsumers, func(i, j int) bool {
-		return existingConsumers[i].IsLessThan(existingConsumers[j])
-	})
+	consumersByPlacement := r.joinConsumersByPlacement(cg.Status.Placements, existingConsumers)
 
-	nConsumers := int32(len(existingConsumers))
-
-	for i := int32(0); i < *cg.Spec.Replicas && i < nConsumers; i++ {
-		if err := r.reconcileConsumer(ctx, cg, existingConsumers[i]); err != nil {
-			return cg.MarkReconcileConsumersFailed("ReconcileConsumer", err)
-		}
-	}
-
-	if nConsumers > *cg.Spec.Replicas {
-		toBeDeleted := existingConsumers[:nConsumers-*cg.Spec.Replicas]
-		for _, c := range toBeDeleted {
-			if err := r.finalizeConsumer(ctx, c); err != nil {
-				return cg.MarkReconcileConsumersFailed("FinalizeConsumer", err)
+	for _, pc := range consumersByPlacement {
+		if pc.Placement == nil {
+			for _, c := range pc.Consumers {
+				if err := r.finalizeConsumer(ctx, c); err != nil {
+					return cg.MarkReconcileConsumersFailed("FinalizeConsumer", err)
+				}
 			}
+			continue
 		}
-		return nil
-	}
-	if nConsumers < *cg.Spec.Replicas {
-		for i := int32(0); i < *cg.Spec.Replicas-nConsumers; i++ {
-			if err := r.reconcileConsumer(ctx, cg, nil); err != nil {
-				return cg.MarkReconcileConsumersFailed("ReconcileConsumer", err)
-			}
+
+		if err := r.reconcileConsumersInPlacement(ctx, cg, *pc.Placement, pc.Consumers); err != nil {
+			return cg.MarkReconcileConsumersFailed("ReconcileConsumer", err)
 		}
-		return nil
 	}
 
 	return nil
 }
 
-func (r Reconciler) reconcileConsumer(ctx context.Context, cg *kafkainternals.ConsumerGroup, c *kafkainternals.Consumer) error {
-	if c == nil {
-		// Consumer doesn't exist, create it.
-		c = cg.ConsumerFromTemplate()
+func (r Reconciler) reconcileConsumersInPlacement(
+	ctx context.Context,
+	cg *kafkainternals.ConsumerGroup,
+	placement eventingduckv1alpha1.Placement,
+	consumers []*kafkainternals.Consumer) error {
 
-		if _, err := r.InternalsClient.Consumers(cg.GetNamespace()).Create(ctx, c, metav1.CreateOptions{}); err != nil {
-			return fmt.Errorf("failed to create consumer %s/%s: %w", c.GetNamespace(), c.GetName(), err)
+	// Check if there is a consumer for the given placement.
```

Comment:
if there is no consumer, we create/reconcile an internal consumer and return it  ?

Reply:
Yes